set of fixes for btrfs graphdriver

nikitych opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->
**- What I did**
Fixes #35630 btrfs graphdriver keeps subvolumes in readonly state while they not used by `Get`.
Rework subvolumes delete function in more efficient way.
Fix quotas usage to handle cases while file system shared with others.

**- How I did it**
I add subvolSetPropRO function to btrfs graphdriver which could set readonly property to subvolume, and call it at Get or Put.
Rework subvolDelete to increase performance by using BTRFS_IOC_TREE_SEARCH and left a full walk of the directory tree as a last resort. Also handle paths longer than PATH_MAX.
Eliminate in-process quota state. Remove ability to disable quotas because only user know when it's save. Invoke quotas rescan when it's really necessary.

**- How to verify it**
Intermediate layers in /var/lib/docker/btrfs/subvolumes/ are read-only. Try to write something and you'll see.
Enable quotas before daemon start or disable while it's running. Qgroups not left after deleted subvols, quota rescan normally not happen.

**- Description for the changelog**
btrfs graphdriver keeps subvolumes in read-only state when possible

nikitych committed:
Change btrfs graphdriver to keep subvolumes in readonly state until used.

Also remove reference to deprecated IOCTL BTRFS_IOC_SUBVOL_CREATE.
BTRFS_IOC_SUBVOL_CREATE_V2 and BTRFS_IOC_SNAP_CREATE_V2 have identical
signature and utilize the same function on kernel side. So both calls
are processed by subvolSnapshot now. In case of src == "" it's creates
empty snapshot - subvol. Therefore, there is no need for a separate
func subvolCreate.

Signed-off-by: Nikita Gerasimov <nikita.gerasimov@oracle.com>

nikitych committed:
Rework subvolDelete to a more efficient algorithm.

As a side effect change openDir to handle correctly a long paths (>=PATH_MAX).

Signed-off-by: Nikita Gerasimov <nikita.gerasimov@oracle.com>

nikitych committed:
Adds SELinux test for btrfs graphdriver

Signed-off-by: Nikita Gerasimov <nikita.gerasimov@oracle.com>

nikitych committed:
Rework btrfs graphdriver quotas usage

The previous implementation did not work optimally and had some errors.
Quotas are enabled and disabled globally, for the entire file system,
and are managed by the kernel. Since the daemon may not be the sole
owner of the fs, for example in the case of btrfs /, it makes no sense
to keep the state within the process. Worse still, disabling quotas
has an unpredictable effect on the user's system. Therefore, the driver
should never turn them off. Last, quotas rescan is a very expensive
operation, but the positive effect is that it can be needed only in
cases of quota inheritance, so we handle it during destroy procedure.

Signed-off-by: Nikita Gerasimov <nikita.gerasimov@oracle.com>

thaJeztah commented:
ping @dmcgowan @kolyshkin PTAL

coolljt0725 commented:
ping @dmcgowan  @kolyshkin

nikitych commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -648,11 +677,9 @@ func (d *Driver) Get(id, mountLabel string) (containerfs.ContainerFS, error) {
 }
 
-// Put is not implemented for BTRFS as there is no cleanup required for the id.
+// Put is return BTRFS subvolume to readonly state.
```
OK

kolyshkin replied:
^^ this, too

nikitych commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -648,11 +677,9 @@ func (d *Driver) Get(id, mountLabel string) (containerfs.ContainerFS, error) {
+// Put is return BTRFS subvolume to readonly state.
 func (d *Driver) Put(id string) error {
-	// Get() creates no runtime resources (like e.g. mounts)
-	// so this doesn't need to do anything.
```
OK

kolyshkin replied:
this, too ^^^

nikitych commented on daemon/graphdriver/btrfs/btrfs_test.go:
```diff
@@ -60,6 +63,73 @@ func TestBtrfsSubvolDelete(t *testing.T) {
+	subdir := x.subvolumesDirID("0subvol")
+	snapdir := x.subvolumesDirID("1snap")
+
+	if err := d.Create("0subvol", "", nil); err != nil {
```
I'll do

nikitych commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -3,14 +3,25 @@
-static void set_name_btrfs_ioctl_vol_args_v2(struct btrfs_ioctl_vol_args_v2* btrfs_struct, const char* value) {
-    snprintf(btrfs_struct->name, BTRFS_SUBVOL_NAME_MAX, "%s", value);
+static int copy_subvolname_and_free(char* dest, char* name) {
+	int r = snprintf(dest, BTRFS_SUBVOL_NAME_MAX, "%s", name);
```
Interesting, will see what we can do.

nikitych replied:
Unfortunately it's introduced in 1.10
https://golang.org/doc/go1.10#cgo

thaJeztah replied:
If we need to backport these changes, we may want to have those changes in a separate commit (as the current stable release of Docker is not yet on Go 1.10)

nikitych commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -165,20 +177,70 @@ func free(p *C.char) {
+	tail := dirpath[len(head)+1:]
+
+	closefd := func(fd C.int) {
+		if r, err := C.close(fd); r != 0 {
```
OK

nikitych commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -165,20 +177,70 @@ func free(p *C.char) {
+			return nil, fmt.Errorf("Can't open dir %s: %v", head, err)
+		}
+		relfd = headfd
+	}
```
Indeed, good catch

nikitych commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -220,84 +282,254 @@ func subvolSnapshot(src, dest, name string) error {
-		args.name[i] = C.char(c)
+	if len(distantDescendants) > 0 {
+		var walk func(*C.DIR, string) error
+		walk = func(pdir *C.DIR, ppath string) error {
```
A recursive call is not compiled without it.

nikitych replied:
I don't mind if it's optimized in the future. But now compilation just fail.

nikitych committed:
Rewrite btrfs tests to use asserts

Package "gotest.tools/assert" used

Signed-off-by: Nikita Gerasimov <nikita.gerasimov@oracle.com>

nikitych committed:
Minor corrections based on code review

Signed-off-by: Nikita Gerasimov <nikita.gerasimov@oracle.com>

kolyshkin commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -311,6 +302,36 @@ func subvolDelete(dirpath, name string, quotaEnabled bool) error {
+		newflags = oldflags &^ C.BTRFS_SUBVOL_RDONLY
+	}
+
+	if newflags != oldflags {
```
style nitpick: better to have

```go
if newflags == oldflags {
	return nil
}
```

while at it, maybe use Go naming convention (i.e. `oldFlags`, `newFlags`).

nikitych replied:
naming convention is OK, but sorry, I don't get why `newflags == oldflags` is better. It's not error to stop function immediately. `newflags != oldflags` clearly indicate optional part and we still have to `return nil` at the end.

kolyshkin commented on daemon/graphdriver/btrfs/btrfs_test.go:
```diff
@@ -172,100 +133,60 @@ func TestBtrfsSubvolLongPath(t *testing.T) {
-		if err := os.Chdir(name); err != nil {
-			t.Fatal(err)
-		}
+		assert.NilError(t, os.Mkdir(name, 0777), "Failed to create dir %s/%s", subvoldir, name)
```
in Go, error message already contains a file name, so you can omit a custom error message entirely, ie
```diff
-		assert.NilError(t, os.Mkdir(name, 0777), "Failed to create dir %s/%s", subvoldir, name)
+		assert.NilError(t, os.Mkdir(name, 0777)

and the error will be  something like

`a_test.go:12: assertion failed: error is not nil: mkdir /some/file: no such file or directory`

which is good enough

nikitych committed:
btrfs graphdriver: replace C.CString to direct work with _GoString_

In go 1.10 a direct access to Go string values from C was introduced.
https://golang.org/doc/go1.10#cgo So now we can save a memcpy in some
places. But the _GoString_ contents may not have a trailing NULL byte
and cannot be modifed. Therefore C.CString is still actual.

Signed-off-by: Nikita Gerasimov <nikita.gerasimov@oracle.com>

kolyshkin commented:
@nikitych are you still able to work on that?

nikitych commented:
@kolyshkin If you describe the desired work in more detail I can give a more accurate answer, but yes I'm still able to do some work on that.

kolyshkin reviewed (changes requested):
I just found I had some old review comments unsent; here they are.

I will do a fresh review soon.

kolyshkin commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -186,51 +186,37 @@ func getDirFd(dir *C.DIR) uintptr {
-		return err
-	}
-	defer closeDir(srcDir)
-
```
this ^^^ is still valid I think

nikitych commented:
@kolyshkin thanks for clarification.

To confirm:
- Close this PR
- Open three new:
  - quotas fixes and tests
  - keeps btrfs subvolumes in readonly state
  - subvolDelete optimization

Am I get it correctly?

kolyshkin commented:
> Am I get it correctly?

Yes, and in case there are changes needed (based on review or otherwise), make sure to modify the existing commits rather than adding new ones on top.

Thank you for your work!

Nefelim4ag commented:
@nikitych, did you give up on that?

Thanks.

nikitych commented:
@Nefelim4ag not really, but unfortunately I have higher priority tasks for now. I hope to send fist PRs before end of the year, but I don't mind if someone want do it more rapid.

olljanat commented:
@nikitych this one needs to be rebased

AkihiroSuda commented:
@nikitych could you rebase?

nikitych committed:
correct names of the dependencies

Signed-off-by: Nikita Gerasimov <nikita.gerasimov@oracle.com>
